### DISCUSSION TITLE
Is it ok to cache imperative atom getter/setter in hook?

### DISCUSSION THREAD

**Discussion by edongashi** [pmndrs/jotai#1543] [Q&A]
Hi.

I wanted to refactor a project from `useContext/State` to something more manageable.
My final options were either Zustand or Jotai. I find Jotai's state derivation and async support superior and went with it.
However, one thing that I consider Jotai is lacking in is the handling of actions.
Writing actions as write-atoms feels cumbersome and unnatural.

The "actions as a hook" from Zustand is lovely, and I want to recreate something similar for Jotai.
After a few attempts I came up with the following: https://codesandbox.io/s/jotai-actions-demo-98f4fg

Snippets from the above demo:

`actions.js`:

```js
const $dispatch = atom(null, (get, set, action) => action(get, set));

export function actions(spec) {
  return function useActions() {
    const dispatch = useSetAtom($dispatch);
    const ref = useRef({
      get() { throw new Error("Component has not yet mounted."); },
      set() { throw new Error("Component has not yet mounted."); }
    });

    useEffect(() => {
      dispatch((get, set) => (ref.current = { get, set }));
    }, [dispatch]);

    return useStaticValue(() => {
      const get = (atom) => ref.current.get(atom);
      const set = (atom, update) => ref.current.set(atom, update);
      return spec(get, set);
    });
  };
}
```

`counter.js`:

```js
export const useCounterActions = actions((get, set) => ({
  increment() {
    set($counter, (counter) => counter + 1);
  },
  decrement() {
    set($counter, (counter) => counter - 1);
  },
  reset() {
    set($counter, 0);
  }
}));
```

As you can see, grabbing a getter and setter pair is quite a hacky process.
I'm wondering if it's wrong to reuse the same functions across multiple actions...
In other words, is `get/set` supposed to be short-lived and/or transactional?

I am aware of `useAtomCallback` but that is not suitable for what I want to create.

Thank you!

**Comment by dai-shi** [comment]
Looks mostly fine, but very hacky. And, useEffect is somewhat misused.
Why doesn't `useAtomCallback` approach work?

Something like this?
```js
export function actions(spec) {
  const $dispatch = atom(null, (get, set, [name, ...args]) => {
    spec(get, set)[name](...args);
  });
  return function useActions() {
    const dispatch = useSetAtom($dispatch);
    return useStaticValue(() => new Proxy({}, {
      get(target, prop) {
        return (...args) => dispatch([prop, ...args]);
      },
    }));
  };
}
```

I used a Proxy, because I don't know the actions in advance. We could run spec(...) at first to list action names, or define the spec signature to be static.

**Reply by edongashi** [reply]
Thank you @dai-shi for the insight!

> And, useEffect is somewhat misused.

Yep, I found no other way to obtain a `set` func while keeping everything referentially stable.
A hypothetical `useAtomGetter/useAtomSetter` would allow building such abstractions easier.

An alternative idea is to run `dispatch` on render, which I'm not sure is safe:

```js
export function actions(spec) {
  return function useActions() {
    const dispatch = useSetAtom($dispatch);
    return useMemo(() => {
      let actions;
      dispatch((get, set) => actions = spec(get, set));
      return actions;
    }, [dispatch]);
  };
}
```

> Why doesn't useAtomCallback approach work?

The callback itself would eventually have to be called, either in `useEffect`, or via proxy like your proposed approach.
I see no benefit over the manual `dispatch/invoke` call.

> I used a Proxy, because I don't know the actions in advance.

Interesting idea. A few notes:

- We have to convert to `useMemo` since `dispatch` may change (?).
- Re-running `spec` on each action imposes the restriction that `spec` must be pure, which I'm fine with.
- The returned callbacks are not referentially stable, so we need to add some caching logic:

```js
return useMemo(() => new Proxy({}, {
  get(target, prop) {
    if (!(prop in target)) {
      target[prop] = (...args) => dispatch([prop, ...args]);
    }
    return target[prop]
  },
}), [dispatch]);
```

> define the spec signature to be static

I really want the Zustand-like syntax :cry:.

**Reply by dai-shi** [reply]
> A hypothetical `useAtomGetter/useAtomSetter` would allow building such abstractions easier.

That's really against the atom abstraction. But, we hope `createStore` would help those use cases.

> An alternative idea is to run `dispatch` on render, which I'm not sure is safe:

It's probably safe, but not idiomatic.

>     * We have to convert to `useMemo` since `dispatch` may change (?).

Yeah.

>     * Re-running `spec` on each action imposes the restriction that `spec` must be pure, which I'm fine with.

😄 

>     * The returned callbacks are not referentially stable, so we need to add some caching logic:

Good catch.

> > define the spec signature to be static
> 
> I really want the Zustand-like syntax 😢.

I see. But, that is the difficulty..

Another trial:
```js
export function actions(spec) {
  const $dispatch = atom(null, (get, set, [name, ...args]) => {
    spec(get, set)[name](...args);
  });
  const names = Object.keys(spec());
  return function useActions() {
    const dispatch = useSetAtom($dispatch);
    return useMemo(() => Object.fromEntries(
      names.map((name) => [name, (...args) => dispatch([name, ...args])])
    ), [dispatch]);
  };
}
```

**Reply by edongashi** [reply]
> Another trial:

I like this one. It feels the least hacky so far. Thanks again!

**Reply by edongashi** [reply]
Another idea that came to mind is to make every `get` and `set` its own operation (untested):

```js
export function actions(spec) {
  const $dispatch = atom(null, (get, set, fn) => fn(get, set));

  return function useActions() {
    const dispatch = useSetAtom($dispatch);

    return useMemo(() => {
      const getFunc = (...args) => {
        let result;
        dispatch((get, set) => result = get(...args));
        return result;
      };
      const setFunc = (...args) => {
        dispatch((get, set) => set(...args));
      };
      return spec(getFunc, setFunc)
    }, [dispatch]);
  };
}

**Reply by dai-shi** [reply]
I just don't prefer the fact that it assumes that `dispatch` runs the function immediately. (otherwise, result will be `undefined` on `return result`.) It's not guaranteed now, but the plan is to make it so in v2.

**Reply by edongashi** [reply]
Ah, right. With batching and transactions and whatnot it can get messy.

I'm close to giving up in favor of a simple `action` factory and `useAction` wrapper. Mostly aliases to distinguish them from state atoms.

```js
const { increment, decrement } = useCounterActions();
// Becomes
const increment = useAction($increment);
const decrement = useAction($decrement);
```

One can then write the `useCounterActions` wrapper easily if needed.

I'm using reverse svelte store syntax to distinguish atoms from values. Time will tell if this is a good idea.

**Reply by edongashi** [reply]
> It's not guaranteed now, but the plan is to make it so in v2.

Does this mean that (currently) inside a write atom, `set(something, 42)` is not guaranteed to take immediate effect?
Or does your comment apply only for `useSetAtom` callback?

I'm also wondering if `$dispatch` is safe to be reused across _all_ actions or should it be unique per action? I.e. what happens if one action calls another action. If the dispatcher is reused then it means the atom is writing itself.

**Reply by dai-shi** [reply]
> > It's not guaranteed now, but the plan is to make it so in v2.

What I meant was `result` can be `undefined`.

> I'm also wondering if `$dispatch` is safe to be reused across _all_ actions or should it be unique per action? I.e. what happens if one action calls another action. If the dispatcher is reused then it means the atom is writing itself.

Reusing `$dispatch` should be safe as long as it's hidden in a hook. I wouldn't think it's idiomatic though. `useAtomCallback` is the same thing, but try to be idiomatic.

The third trial:
```js
export function actions(spec) {
  const names = Object.keys(spec());
  const createCallbacks = () => Object.fromEntries(name.map((name => [
    name,
    // supporting only one arg (but it can be improved)
    useAtomCallback((get, set, arg) => spec(get, set)[name](arg)),
  ]));
  return function useActions() {
    return useMemo(createCallbacks, []);
  };
}
```
